smarts/sstudio/genscenario.py

```diff
@@ -54,95 +149,205 @@ def gen_scenario(
     # XXX: For now this simply coalesces the sub-calls but in the future this allows
     #      us to simplify our serialization between SStudio and SMARTS.
 
-    output_dir = str(output_dir)
-
-    if scenario.map_spec:
-        gen_map(output_dir, scenario.map_spec)
+    scenario_dir = os.path.abspath(str(output_dir))
+    build_dir = os.path.join(scenario_dir, "build")
+    build_graph = _build_graph(scenario, build_dir)
+    os.makedirs(build_dir, exist_ok=True)
+
+    # Create DB for build caching
+    db_path = os.path.join(build_dir, "build.db")
+    db_conn = sqlite3.connect(db_path)
+
+    cur = db_conn.cursor()
+    cur.execute(
+        """CREATE TABLE IF NOT EXISTS Artifacts (
+                artifact_path     TEXT PRIMARY KEY,
+                scenario_obj_hash TEXT
+        ) WITHOUT ROWID"""
+    )
+    db_conn.commit()
+    cur.close()
+
+    # Map spec
+    artifact_paths = build_graph["map_spec"]
+    obj_hash = pickle_hash(scenario.map_spec, True)
+    if _needs_build(db_conn, scenario.map_spec, artifact_paths, obj_hash):
+        with timeit("map_spec", logger.info):
+            _gen_map(scenario_dir, scenario.map_spec)
+            _update_artifacts(db_conn, artifact_paths, obj_hash)
         map_spec = scenario.map_spec
     else:
-        map_spec = types.MapSpec(source=output_dir)
+        map_spec = types.MapSpec(source=scenario_dir)
 
-    if scenario.traffic:
-        for name, traffic in scenario.traffic.items():
-            gen_traffic(
-                scenario=output_dir,
-                traffic=traffic,
-                name=name,
-                seed=seed,
-                overwrite=overwrite,
-                map_spec=map_spec,
-            )
+    if map_spec.shift_to_origin and scenario.traffic_histories:
+        logger.warning(
+            "Attempting to shift map with traffic histories."
+            "The data may no longer line up with the map."
+        )
```

Comment:
Is this not just an error?

Reply:
This is how it worked before, I just kept the same behaviour. I'm open either way though. Is there a situation where there would be traffic histories and shifting the map would still be a valid thing to do?

Reply:
I do not think there is a case because it messes with the data. However, if we were to discard the underlying data this would still be OK. Maybe we leave as is.